### PR TITLE
RESOLVES #2739: Merge partitions when they become too small

### LINKED ISSUE #2739: Merge partitions when they become too small
Merge partitions when they become too small

### PR DESCRIPTION
**Pull request by wshaib-apple**
- when a partition falls below the low watermark, and if one or both of its neighboring partitions have capacity to absorb, it will be merged into them.

### CODE REVIEW

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java [hunk 1]
```diff
@@ -444,28 +454,42 @@ public String getPartitionFieldNameInLucene() {
      *
      * @param newRecord record to be written
      * @param groupingKey grouping key
+     * @param assignedPartitionIdHint assigned partition override, if not null
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java:457]: `hint` seems misleading, it looks like this is used no matter what, replacing any attempt to lookup what it should be.

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java [hunk 2]
```diff
@@ -1707,6 +1718,89 @@ private List<Tuple> queryLocal(Map<Tuple, Long> dataset, Comparisons.Type compar
...
+
+        // assert correct partition setup after repartitioning
+        assertEquals(expectedPartitionCounts.length, partitionInfos.size());
+        assertArrayEquals(expectedPartitionCounts, partitionInfos.stream().mapToInt(LucenePartitionInfoProto.LucenePartitionInfo::getCount).toArray());
+
+        try (FDBRecordContext context = openContext(contextProps)) {
+            schemaSetup.accept(context);
+            // finally, assert all docs are accounted for
+            assertIndexEntryPrimaryKeyTuples(createdKeys.get(Tuple.from(1L)).keySet(),
+                    recordStore.scanIndex(COMPLEX_PARTITIONED, groupedTextSearch(COMPLEX_PARTITIONED, luceneSearch, 1), null, ExecuteProperties.newBuilder().build().asScanProperties(false)));
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java:1800]: This doesn't seem to validate that the right documents are in the right partitions, it seems important to run `LuceneIndexTestValidator.validate` on this, to ensure that the `from`/`to` all make sense for the documents contained, and that they are still ordered correctly.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java [hunk 3]
```diff
@@ -958,25 +1076,42 @@ private CompletableFuture<Integer> moveDocsFromPartition(@Nonnull final LucenePa
...
                 // value of the "destination" partition's `from` value
                 final Tuple overflowPartitioningKey = toPartitionKey(records.get(0));
-                return findPartitionInfo(groupingKey, overflowPartitioningKey).thenCompose(previousPartition -> {
-                    if (previousPartition == null || previousPartition.getCount() + records.size() > indexPartitionHighWatermark || previousPartition.getId() == partitionInfo.getId()) {
+                final CompletableFuture<LucenePartitionInfoProto.LucenePartitionInfo> futureDestination;
+                if (repartitioningContext.removingOldest) {
+                    futureDestination = findPartitionInfo(groupingKey, overflowPartitioningKey);
+                } else {
+                    futureDestination = getNextNewerPartitionInfo(state.context, groupingKey, overflowPartitioningKey, state.indexSubspace);
+                }
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java:1104]: Well above this we get all the partition info and put it in a list, would it make sense to sort that list, and use that instead of using these methods?
The downside is that you would need two versions of the code for `findPartitionInfo` and `getNextNewerPartitionInfo` depending on whether you're hitting FDB or a list, but it might be a little bit cleaner/faster.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java [hunk 4]
```diff
@@ -958,25 +1076,42 @@ private CompletableFuture<Integer> moveDocsFromPartition(@Nonnull final LucenePa
...
+                return futureDestination.thenCompose(destinationPartition -> {
+                    if (destinationPartition == null || destinationPartition.getCount() + records.size() > indexPartitionHighWatermark || destinationPartition.getId() == partitionInfo.getId()) {
                         // create a new "overflow" partition
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java:1107]: Is it possible for it to hit this code when shrinking a partition?

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java [hunk 5]
```diff
@@ -800,57 +833,121 @@ public CompletableFuture<RecordCursorContinuation> rebalancePartitions(RecordCur
...
 
+    /**
+     * convenience function that proxies into {@link #moveDocsFromPartition(LucenePartitionInfoProto.LucenePartitionInfo, Tuple, RepartitioningContext)}
+     * after outputting some logs.
+     *
+     * @param partitionInfo partition info
+     * @param groupingKey grouping key
+     * @param repartitioningContext repartition parameters
+     * @param logMessages log messages
+     * @return future count of documents mvoed
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java:931]:
```suggestion
     * @return future count of documents moved
```

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java [hunk 6]
```diff
@@ -800,57 +833,121 @@ public CompletableFuture<RecordCursorContinuation> rebalancePartitions(RecordCur
...
+                    }
+                } else {
+                    // partition is above high watermark
+                    repartitioningContext.countToMove = 1 + Math.min(repartitionDocumentCount, indexPartitionHighWatermark);
+                    // sanity: in case someone passes a high enough repartitionDocumentCount such that removing
+                    // this many docs would flip the partition from being above the high watermark to being below
+                    // the low watermark
+                    if (currentPartitionCount - repartitioningContext.countToMove < indexPartitionLowWatermark) {
+                        repartitioningContext.countToMove = indexPartitionHighWatermark - indexPartitionLowWatermark;
+                    }
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java:912]: I guess it's probably a bug if this happens, so maybe optimizing it is irrelevant, but it seems to me that you would want to move `(indexPartitionHighWatermark + indexPartitionLowWatermark) / 2`, rather than bringing it all the way to the low watermark.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java [hunk 7]
```diff
@@ -790,6 +814,15 @@ public CompletableFuture<RecordCursorContinuation> rebalancePartitions(RecordCur
         }
     }
 
+    private static class RepartitioningContext {
+        boolean emptyingPartition = false;
+        LuceneRecordCursor cursor;
+        int countToMove;
+        boolean removingOldest = true;
+        int maxPartitionId;
+        boolean newBoundaryRecordPresent = false;
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java:823]: What is this one for?

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java [hunk 8]
```diff
@@ -800,57 +833,121 @@ public CompletableFuture<RecordCursorContinuation> rebalancePartitions(RecordCur
...
+                        repartitioningContext.countToMove = Math.min(previousCapacity, repartitionDocumentCount);
+                        if (currentPartitionCount > repartitioningContext.countToMove) {
+                            repartitioningContext.countToMove++;
+                            repartitioningContext.newBoundaryRecordPresent = true;
+                        }
+                        repartitioningContext.emptyingPartition = false;
+                        repartitioningContext.cursor = getOldestNDocuments(partitionInfo, groupingKey, repartitioningContext.countToMove);
+                    } else {
+                        // partition is under low-watermark, but no capacity around it to absorb it
+                        continue;
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java:902]: does `repartitioningCount` need to get reset if this happens? 
I think the flow would have to be something like:
1. see a partition that you should empty
2. all of the if/else-if above return false so it gets to here
3. next partition should not be emptied.

The only other field that is set is the `countToMove` which always gets set.
It does seem a bit brittle though to have it living between partitions, even if resetting `emptyingPartition` would be sufficient for right now.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java [hunk 9]
```diff
@@ -800,57 +833,121 @@ public CompletableFuture<RecordCursorContinuation> rebalancePartitions(RecordCur
...
+
+                final int currentPartitionCount = partitionInfo.getCount();
+
+                if (currentPartitionCount >= indexPartitionLowWatermark && currentPartitionCount <= indexPartitionHighWatermark) {
+                    // repartitioning not needed
+                    continue;
+                }
+
+                if (currentPartitionCount < indexPartitionLowWatermark) {
+                    // partitionInfos are ordered next -> previous
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java:872]: `next` and `previous` I think are new terms only used here, perhaps replacing their usage with `newer` and `older` respectively.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java [hunk 10]
```diff
@@ -800,57 +833,121 @@ public CompletableFuture<RecordCursorContinuation> rebalancePartitions(RecordCur
...
+                    LucenePartitionInfoProto.LucenePartitionInfo next = i == 0 ? null : partitionInfos.get(i - 1);
+                    LucenePartitionInfoProto.LucenePartitionInfo previous = i == partitionInfos.size() - 1 ? null : partitionInfos.get(i + 1);
+                    repartitioningContext.countToMove = Math.min(currentPartitionCount, repartitionDocumentCount);
+                    if (currentPartitionCount > repartitionDocumentCount) {
+                        repartitioningContext.countToMove++;
+                    } else {
+                        repartitioningContext.emptyingPartition = true;
                     }
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java:880]: It _might_ be clearer to remove the `Math.min` and just set `countToMove` in both blocks
```suggestion
                    if (currentPartitionCount > repartitionDocumentCount) {
                        repartitioningContext.countToMove = repartitionDocumentCount + 1;
                    } else {
                        repartitioningContext.countToMove = currentPartitionCount;
                        repartitioningContext.emptyingPartition = true;
                    }
```

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java [hunk 11]
```diff
@@ -1707,6 +1718,89 @@ private List<Tuple> queryLocal(Map<Tuple, Long> dataset, Comparisons.Type compar
...
+                Arguments.of(2, 4, 3, new int[] {1, 1}, new int[] {2}),
+                // consolidate a partition into its previous neighbor
+                Arguments.of(2, 4, 3, new int[] {2, 2, 1}, new int[] {2, 3}),
+                // consolidate a partition falling between two partitions with capacity into its previous neighbor
+                Arguments.of(3, 7, 3, new int[] {5, 2, 5}, new int[] {7, 5}),
+                // consolidate a partition falling between two partitions which individually don't have enough
+                // capacity, but together they do
+                Arguments.of(3, 7, 3, new int[] {6, 2, 6}, new int[] {7, 7}),
+                // cannot consolidate a partition that has no neighbors with capacity
+                Arguments.of(3, 7, 3, new int[] {6, 2, 7}, new int[] {6, 2, 7}),
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java:1733]: It seems like you should have the same test starting with `{7, 2, 6}`

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java [hunk 12]
```diff
@@ -1707,6 +1718,89 @@ private List<Tuple> queryLocal(Map<Tuple, Long> dataset, Comparisons.Type compar
...
+                // cannot consolidate partitions that have no neighbors with capacity
+                Arguments.of(3, 7, 3, new int[] {6, 2, 7, 3}, new int[] {6, 2, 7, 3}),
+                // consolidate one partition that has a neighbor with capacity, while another
+                // that doesn't, won't be consolidated
+                Arguments.of(3, 7, 3, new int[] {6, 2, 6, 1}, new int[] {6, 2, 7}),
+                // splitting one partition, removes the need to consolidate a previous low partition
+                Arguments.of(3, 7, 3, new int[] {6, 1, 8}, new int[] {6, 4, 5}),
+                // consolidating two neighboring partitions into their left and right neighbors
+                Arguments.of(3, 7, 3, new int[] {6, 1, 1, 6}, new int[] {7, 7}),
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java:1742]: What about
```
Arguments.of(4, 7, 3, new int[] {6, 3, 4}, new int[] { ? ? }),
```

And should that behave differently if there's a whole bunch of newer partitions

```
Arguments.of(4, 7, 3, new int[] {6, 3, 4, 5, 5, 5, 5}, new int[] { ? ? }),
```

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java [hunk 13]
```diff
@@ -958,25 +1076,42 @@ private CompletableFuture<Integer> moveDocsFromPartition(@Nonnull final LucenePa
...
+                        destinationPartition = newPartitionMetadata(overflowPartitioningKey,  repartitioningContext.maxPartitionId + 1);
+                        savePartitionMetadata(groupingKey, destinationPartition.toBuilder());
                     }
 
                     Iterator<? extends FDBIndexableRecord<Message>> recordIterator = records.iterator();
-                    return AsyncUtil.whileTrue(() -> indexMaintainer.update(null, recordIterator.next())
+                    final int destinationPartitionId = destinationPartition.getId();
+                    return AsyncUtil.whileTrue(() -> indexMaintainer.update(null, recordIterator.next(), destinationPartitionId)
                             .thenApply(ignored -> recordIterator.hasNext()), state.context.getExecutor());
                 }).thenApply(ignored -> records.size());
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java:1116]: I don't think we close the cursor used to get the newest or oldest documents.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java [hunk 14]
```diff
@@ -790,6 +814,15 @@ public CompletableFuture<RecordCursorContinuation> rebalancePartitions(RecordCur
         }
     }
 
+    private static class RepartitioningContext {
+        boolean emptyingPartition = false;
+        LuceneRecordCursor cursor;
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java:819]: Would it make sense to have this have the target partition instead, and call `getOldestestNDocuments` or `getNewestNDocuments` based on `removingOldest`, closer to when the cursor is being used.

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java [hunk 15]
```diff
@@ -1707,6 +1718,89 @@ private List<Tuple> queryLocal(Map<Tuple, Long> dataset, Comparisons.Type compar
         return hits.stream().map(Map.Entry::getKey).collect(Collectors.toList());
     }
 
+    static Stream<Arguments> simplePartitionConsolidationTest() {
+        return Stream.of(
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java:1722]: Thinking about the testability, and complexity here, I wonder if it would make sense, and not be a big change, to have something like a `LuceneRepartitionPlanner` that takes a `List<LucenePartitionInfoProto.LucenePartitionInfo>` and returns a plan, which is something like:
> Move N newest documents from partition 3 to partition 4 and you are emptying partition 3.

Or

> Move M oldest documents from partition 2 to partition 1.

And then `LucenePartitioner` takes that plan and executes it in a single transaction. I'm not sure who is in charge of whether there is more to do.
That would probably allow decoupling the tests so you can have a lot of unit tests of the `RepartitionPlanner`, and fewer tests that actually involve the whole machinery and saving actual records to the database.

**Review by ScottDugas — CHANGES_REQUESTED** [review]

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java [hunk 16]
```diff
@@ -1707,6 +1718,89 @@ private List<Tuple> queryLocal(Map<Tuple, Long> dataset, Comparisons.Type compar
...
+
+        List<LucenePartitionInfoProto.LucenePartitionInfo> partitionInfos = getPartitionMeta(index,
+                Tuple.from(1), contextProps, schemaSetup);
+        partitionInfos.sort(Comparator.comparing(LucenePartitionInfoProto.LucenePartitionInfo::getId));
+
+        // assert sane initial setup
+        assertEquals(initialPartitionCounts.length, partitionInfos.size());
+        assertArrayEquals(initialPartitionCounts, partitionInfos.stream().mapToInt(LucenePartitionInfoProto.LucenePartitionInfo::getCount).toArray());
+
+        explicitMergeIndex(index, contextProps, schemaSetup);
```
[the remark this answers is not in the document]
Comment (wshaib-apple) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java:1787]: I think I ensured no infinite loop since we would never move enough docs to cause a destination partition to go above its high watermark...

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java (hunk 2 above)
Comment (wshaib-apple) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java:1800]: see above

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java (hunk 6 above)
Comment (wshaib-apple) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java:912]: you mean `(indexPartitionHighWatermark - indexPartitionLowWatermark) / 2` ?

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java (hunk 3 above)
Comment (wshaib-apple) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java:1104]: in addition,

1. we would need to pass the partition info list down to this method (in the `repartitioningContext`?
2. make sure while implementing the new versions of `findPartitionInfo` and `getNextNewerPartitionInfo` to be aware whether the current partition has been cleared in the DB (a few lines above here).

is the gain in speed is worth this added complexity?

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java [hunk 17]
```diff
@@ -1707,6 +1718,89 @@ private List<Tuple> queryLocal(Map<Tuple, Long> dataset, Comparisons.Type compar
...
+                Arguments.of(3, 7, 3, new int[] {6, 2, 7, 3}, new int[] {6, 2, 7, 3}),
+                // consolidate one partition that has a neighbor with capacity, while another
+                // that doesn't, won't be consolidated
+                Arguments.of(3, 7, 3, new int[] {6, 2, 6, 1}, new int[] {6, 2, 7}),
+                // splitting one partition, removes the need to consolidate a previous low partition
+                Arguments.of(3, 7, 3, new int[] {6, 1, 8}, new int[] {6, 4, 5}),
+                // consolidating two neighboring partitions into their left and right neighbors
+                Arguments.of(3, 7, 3, new int[] {6, 1, 1, 6}, new int[] {7, 7}),
+                // multiple-stage split
+                Arguments.of(3, 7, 3, new int[] {15}, new int[] {6, 6, 3})
```
[the remark this answers is not in the document]
Comment (wshaib-apple) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java:1744]: > Relatedly all the tests have a `repartitionCount>=lowWatermark`, which I don't think is really what you would want....

true, they are, but that actually is a more stringent test. having `repartitionCount < lowWatermark` may not uncover some edge cases. I can make it pass both kinds of values.

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java (hunk 11 above)
Comment (wshaib-apple) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java:1733]: Arguments.of(5, 10, 1, new int[] {7, 4, 10}, new int[] {7, 4, 10}),
